### PR TITLE
Introduce PSScriptAnalyzer to CI

### PR DESCRIPTION
**Pull request by m-kostrzewa**

### CODE REVIEW

**Comment by m-kostrzewa** [comment]
recheck no bug

File: CIScripts/Build/BuildFunctions.ps1 [hunk 1]
```diff
@@ -69,12 +38,12 @@ function Invoke-DockerDriverBuild {
...
            [Parameter(Mandatory = $true)] [string] $SigntoolPath,
            [Parameter(Mandatory = $true)] [string] $CertPath,
-           [Parameter(Mandatory = $true)] [string] $CertPasswordFilePath,
+           [Parameter(Mandatory = $true)] [string] $CertPasswdFilePath,
            [Parameter(Mandatory = $true)] [string] $OutputPath,
            [Parameter(Mandatory = $true)] [string] $LogsPath)
 
     $Job.PushStep("Docker driver build")
-    $GoPath = if (Test-Path Env:GOPATH) { (pwd) + ";$Env:GOPATH" } else { pwd }
+    $GoPath = if (Test-Path Env:GOPATH) { ( Get-Location ) + ";$Env:GOPATH" } else { Get-Location }
```
Comment (mclapinski) [inline CIScripts/Build/BuildFunctions.ps1]: Split into:
```
$GoPath = Get-Location
if (Test-Path Env:GOPATH) {
    $GoPath += ";$Env:GOPATH"
}

File: CIScripts/Common/Job.ps1 [hunk 2]
```diff
@@ -76,7 +77,11 @@ class JobStep {
 
     Print([int] $IndentLevel) {
         $msg = ""
-        1..($IndentLevel) | ForEach-Object{ $msg += " " }
+        1..($IndentLevel) | ForEach-Object {
+            [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseDeclaredVarsMoreThanAssignments",
+                "", Justification="Issue #804")]
```
Comment (mclapinski) [inline CIScripts/Common/Job.ps1]: should be something like "Issue #804 from PSScriptAnalyzer" or just paste the link https://github.com/PowerShell/PSScriptAnalyzer/issues/804

File: CIScripts/Common/VMUtils.ps1 [hunk 3]
```diff
@@ -33,10 +38,10 @@ function New-RemoteSessionsToTestbeds {
         throw "Cannot create remote sessions to testbeds: $Env:TESTBED_ADDRESSES not set"
     }
 
-    $Creds = Get-VMCreds
+    $Creds = Get-VMCredential
```
Comment (mclapinski) [inline CIScripts/Common/VMUtils.ps1]: fix

File: CIScripts/Test/Tests/vRouterAgentService.Tests.ps1 [hunk 4]
```diff
@@ -55,6 +55,8 @@ Describe "vRouter Agent service" {
                 -AdapterName $TestConf.AdapterName `
                 -VMSwitchName $TestConf.VMSwitchName `
                 -ForwardingExtensionName $TestConf.ForwardingExtensionName
+            [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseDeclaredVarsMoreThanAssignments",
+                "", Justification="PSA #804")]
```
Comment (mclapinski) [inline CIScripts/Test/Tests/vRouterAgentService.Tests.ps1]: For issue 804 you use several different justifications

File: Linters/Invoke-AllLinters.ps1 [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+$PSLinterConfig = "$ConfigDir/PSScriptAnalyzerSettings.psd1"
+Write-Host "Running PSScriptAnalyzer... (config from $PSLinterConfig)"
+
+$Output = Invoke-ScriptAnalyzer $RootDir -Recurse -Setting $PSLinterConfig
+if ($Output) {
+      Write-Host ($Output | Format-Table | Out-String)
+      exit 1
+}
+
+exit 0
```
Comment (mclapinski) [inline Linters/Invoke-AllLinters.ps1]: new line

File: Linters/PSScriptAnalyzerSettings.psd1 [hunk 6]
```diff
@@ -0,0 +1,12 @@
+@{
+    Severity=@("Error", "Warning")
+    ExcludeRules=@(
+        "PSAvoidUsingWriteHost",        # This will be removed in favor of Write-Log in the future.
+        "PSAvoidUsingInvokeExpression", # We don't have an alternative for this yet.
+        "PSUseShouldProcessForStateChangingFunctions",      # We don't care about altering system state.
+        "PSAvoidUsingConvertToSecureStringWithPlainText",   # We have creds mostly to testbed
```
Comment (mclapinski) [inline Linters/PSScriptAnalyzerSettings.psd1]: it's a little weird that you disable it globally but disable PSAvoidUsingPlainTextForPassword locally

File: Linters/README.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# Run all available linters
+
+To run all available linters, execute the following command in this directory:
+
+```
+.\Invoke-AllLinters.ps1 -RootDir .. -ConfigDir $pwd
```
Comment (mclapinski) [inline Linters/README.md]: why `$pwd` instead of `.`?

**Review by mclapinski — DISMISSED** [review]

File: Linters/README.md (hunk 7 above)
Comment (m-kostrzewa) [reply Linters/README.md]: PSScriptAnalyzer quirk. I can't get it to work with config from rel path....

**Review by kaweue — APPROVED** [review]
MC + PM

**Pull request merged**